internal/controllers/core/filewatch/watch_manager.go

```diff
@@ -0,0 +1,237 @@
+)
+
+const maxFileWatchHistory = 20
+
+// NotifyClient is the minimal set of methods for FileWatch API that the manager needs.
+//
+// This matches the methods from tiltapi FileWatchInterface, NOT the controller client!
+// The controller client is only suitable for usage _within_ reconciliation as it does
+// its own caching.
+type NotifyClient interface {
+	Get(ctx context.Context, name string, opts v1.GetOptions) (*filewatches.FileWatch, error)
+	UpdateStatus(ctx context.Context, fileWatch *filewatches.FileWatch, opts v1.UpdateOptions) (*filewatches.FileWatch, error)
+}
+
+type ApiServerWatchManager struct {
+	client NotifyClient
+
+	watcherMaker watch.FsWatcherMaker
+	timerMaker   watch.TimerMaker
+
+	mu      sync.Mutex
+	watches map[string]fsWatch
+}
+
+type fsWatch struct {
+	spec     filewatches.FileWatchSpec
+	notifier watch.Notify
+	cancel   func()
+}
+
+func NewApiServerWatchManager(clientset tiltapi.Interface, watcherMaker watch.FsWatcherMaker, timerMaker watch.TimerMaker) *ApiServerWatchManager {
+	return &ApiServerWatchManager{
+		client:       clientset.CoreV1alpha1().FileWatches(),
+		watcherMaker: watcherMaker,
+		timerMaker:   timerMaker,
+		watches:      make(map[string]fsWatch),
+	}
+}
+
+func (m *ApiServerWatchManager) StopWatch(name string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	w, ok := m.watches[name]
+	if !ok {
+		return nil
+	}
+
+	if err := w.notifier.Close(); err != nil {
+		return err
+	}
+	w.cancel()
+	return nil
+}
+
+// StartWatch adds a new filesystem watch for a FileWatch API object.
+//
+// If a watch for the spec already exists and the spec has not changed, the existing watch is left untouched
+// and the boolean return value is false. If there is no watch or the spec has changed, it is replaced and
+// the boolean return value is true.
+func (m *ApiServerWatchManager) StartWatch(ctx context.Context, name string, spec filewatches.FileWatchSpec) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	oldWatch, watchAlreadyExists := m.watches[name]
+	if watchAlreadyExists {
+		if equality.Semantic.DeepEqual(oldWatch.spec, spec) {
+			return false, nil
+		}
+	}
+
+	ignoreMatcher, err := dockerignore.NewDockerPatternMatcher(spec.RootPath, spec.IgnorePatterns)
+	if err != nil {
+		return false, err
+	}
+
+	watchPaths, err := absPaths(spec.RootPath, spec.Paths)
+	if err != nil {
+		return false, err
+	}
+
+	notifier, err := m.watcherMaker(watchPaths, ignoreMatcher, logger.Get(ctx))
+	if err != nil {
+		return false, err
+	}
+
+	if err := notifier.Start(); err != nil {
+		return false, err
+	}
+
+	watchCtx, cancel := context.WithCancel(ctx)
+
+	w := fsWatch{
+		spec:     spec,
+		notifier: notifier,
+		cancel:   cancel,
+	}
+
+	go m.notifyLoop(watchCtx, name, notifier)
+
+	// the old watch is cleaned up AFTER starting the new one to avoid missing events
+	// (this might mean duplicates are received, which is an acceptable trade-off)
+	if oldWatch.cancel != nil {
+		oldWatch.cancel()
```

Comment:
it seems odd to me that this implementation looks different from StopWatch?

Reply:
Cleaned this up - new logic is consistent and a bit paranoid to balance not leaking resources but also avoid any weird races